arangod/Replication2/ReplicatedLog/LogLeader.cpp

```diff
@@ -0,0 +1,945 @@
+    }
+    auto const index = leaderData._inMemoryLog.getNextIndex();
+    auto const payloadSize = payload.byteSize();
+    auto logEntry =
+        InMemoryLogEntry(PersistingLogEntry(_currentTerm, index, std::move(payload)));
+    logEntry.setInsertTp(insertTp);
+    leaderData._inMemoryLog.appendInPlace(_logContext, std::move(logEntry));
+    _logMetrics->replicatedLogInsertsBytes->count(payloadSize);
+    return index;
+  });
+}
+
+auto replicated_log::LogLeader::waitFor(LogIndex index) -> WaitForFuture {
+  return _guardedLeaderData.doUnderLock([index](auto& leaderData) {
+    if (leaderData._didResign) {
+      auto promise = WaitForPromise{};
+      promise.setException(basics::Exception(TRI_ERROR_REPLICATION_REPLICATED_LOG_LEADER_RESIGNED,
+                                             __FILE__, __LINE__));
+      return promise.getFuture();
+    }
+    if (leaderData._commitIndex >= index) {
+      return futures::Future<std::shared_ptr<QuorumData const>>{std::in_place,
+                                                                leaderData._lastQuorum};
+    }
+    auto it = leaderData._waitForQueue.emplace(index, WaitForPromise{});
+    auto& promise = it->second;
+    auto&& future = promise.getFuture();
+    TRI_ASSERT(future.valid());
+    return std::move(future);
+  });
+}
+
+auto replicated_log::LogLeader::getParticipantId() const noexcept -> ParticipantId const& {
+  return _id;
+}
+
+auto replicated_log::LogLeader::triggerAsyncReplication() -> void {
+  auto preparedRequests = _guardedLeaderData.doUnderLock([](auto& leaderData) {
+    if (leaderData._didResign) {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_REPLICATION_REPLICATED_LOG_LEADER_RESIGNED);
+    }
+    return leaderData.prepareAppendEntries();
+  });
+  executeAppendEntriesRequests(std::move(preparedRequests), _logMetrics);
+}
+
+auto replicated_log::LogLeader::GuardedLeaderData::updateCommitIndexLeader(
+    LogIndex newIndex, std::shared_ptr<QuorumData> const& quorum) -> ResolvedPromiseSet {
+  LOG_CTX("a9a7e", TRACE, _self._logContext)
+      << "updating commit index to " << newIndex << "with quorum " << quorum->quorum;
+  auto oldIndex = _commitIndex;
+
+  TRI_ASSERT(_commitIndex < newIndex);
+  _commitIndex = newIndex;
+  _lastQuorum = quorum;
+
+  try {
+    WaitForQueue toBeResolved;
+    auto const end = _waitForQueue.upper_bound(_commitIndex);
+    for (auto it = _waitForQueue.begin(); it != end;) {
+      LOG_CTX("37d9c", TRACE, _self._logContext)
+          << "resolving promise for index " << it->first;
+      toBeResolved.insert(_waitForQueue.extract(it++));
+    }
+    return ResolvedPromiseSet{std::move(toBeResolved), quorum,
+                              _inMemoryLog.splice(oldIndex, newIndex + 1)};
+  } catch (...) {
+    // If those promises are not fulfilled we can not continue.
+    // Note that the move constructor of std::multi_map is not noexcept.
+    abort();
```

Comment:
Should this be `std::abort();`? Shall we write a descriptive log message about what happened?
```suggestion
    std::abort();
```

Reply:
I added log output and a `FATAL_ERROR_EXIT` instead.